p2panda-rs/src/encoder.rs

```diff
@@ -0,0 +1,286 @@
+            decode_entry(&second_signed_encoded_entry, Some(&second_encoded_message)).unwrap();
+
+        // All decoded_entry and second_decoded_entry Entry params should not be equal
+        // except for LogId (1) and skiplink_hash (None)
+        assert_eq!(decoded_entry.log_id(), second_decoded_entry.log_id());
+        assert_ne!(
+            decoded_entry.message().unwrap(),
+            second_decoded_entry.message().unwrap()
+        );
+        assert_ne!(decoded_entry.seq_num(), second_decoded_entry.seq_num());
+        assert_ne!(
+            decoded_entry.backlink_hash(),
+            second_decoded_entry.backlink_hash()
+        );
+        assert_eq!(
+            decoded_entry.skiplink_hash(),
+            second_decoded_entry.skiplink_hash()
+        );
+    }
+
+    #[test]
+    fn entry_signing_and_encoding_fixtures() {
+        let public_key = "ba07a8da75dd2f922d62eae7e7ac7c081e06bf0c192b2d8ea1b2ab5e9c59013e";
+        let private_key = "31f33f8e6c262f36a0e5397348093a459d66d8cb5946798ad62d5eb8e7645bdb";
+        let author = "ba07a8da75dd2f922d62eae7e7ac7c081e06bf0c192b2d8ea1b2ab5e9c59013e";
+        let entry_bytes= "00ba07a8da75dd2f922d62eae7e7ac7c081e06bf0c192b2d8ea1b2ab5e9c59013e01040040944b4ae2ff31d0adc13cf94ba43b766871b4e56e96d0eebbc1b9e2b8226d448e8bc1f9507a21894579578491ff778a008688c2a3e8a409fc37522d9eabaa114c004054f65f3ac2ccf13f5862eb7c29ac20e830e173d062416dfd03a27e8a2315b69f402cfa4ca741d243b184b1d8ff203cf1f1ec4619f44758263f19a75a3537e780ee00408960c9d4f864aef757d440bc5aa5a5c0d726312eddadad68f25d06fedd10f755d51a87565972f8c3d77ef7ac66531227131b0d8857fef749c3a98cfffae8519d1e8bdb78a27348232671acda6c16aca26148642b0e803e6e2e4dfc01ca0d46ea19546be7b4302b826363a6caa28fced7ef9fd847b35a49eb67b885d65af14305";
+        let entry_hash = "004073412203af4eedbddde3a8183647f3c788667b4e693ebdd38830eeaef10cbdf956587f0f571213ba5a86b2cbc5926c8b6b27e5f721a69ced281aa8ad57aa4404";
+        let log_id: i64 = 1;
+        let payload_bytes = "a466616374696f6e6663726561746566736368656d6178843030343063663934663664363035363537653930633534336230633931393037306364616166373230396335653165613538616362386633353638666132313134323638646339616333626166653132616632373764323836666365376463353962376330633334383937336334653964616362653739343835653536616332613730326776657273696f6e01666669656c6473a26464617465a164546578747818323032312d30352d30325432303a30363a34352e3433305a676d657373616765a164546578746d477574656e204d6f7267656e21";
+        let payload_hash = "00408960c9d4f864aef757d440bc5aa5a5c0d726312eddadad68f25d06fedd10f755d51a87565972f8c3d77ef7ac66531227131b0d8857fef749c3a98cfffae8519d";
+        let seq_num: i64 = 4;
+        let skiplink_hash = "0040944b4ae2ff31d0adc13cf94ba43b766871b4e56e96d0eebbc1b9e2b8226d448e8bc1f9507a21894579578491ff778a008688c2a3e8a409fc37522d9eabaa114c";
+        let backlink_hash = "004054f65f3ac2ccf13f5862eb7c29ac20e830e173d062416dfd03a27e8a2315b69f402cfa4ca741d243b184b1d8ff203cf1f1ec4619f44758263f19a75a3537e780";
```

Comment:
We had previously used a round-trip testing model (it's still there in `entry.rs`). That is, we create an encoded data structure, decode it, re-encode it and then compare the two encoded forms to ensure that de- and encoding doesn't change the data.

This bit here introduces another approach that doesn't prevent inconsistencies within a conversion cycle but across versions, i.e. if the library is changed and these values stay here in the test, the test will only fail if the new library doesn't handle these old values properly.

I like that! Maybe we should actually make this more explicit by turning these constants into test fixtures annotated with the specific library version they were created with. Then eventually we could have tests that iterate over data produced by all previously published api versions to make sure all of them can be handled.

Of course that doesn't need to happen in this PR, just a thought.

Reply:
Ah yeh, this is a really good idea, let's store it for later.

Should we bring the round trip test from entry into this module so they can be side by side and the distinction made clearer?

Reply:
I actually had a go at implementing fixture tests with `rstest`, sure some of it needs some work still, but as a starting point I don't think it's too bad.

https://github.com/p2panda/p2panda/blob/010416d386f0bb7d94823264c14f4abd437768a8/p2panda-rs/src/encoder.rs#L167-L183

https://github.com/p2panda/p2panda/blob/010416d386f0bb7d94823264c14f4abd437768a8/p2panda-rs/src/encoder.rs#L200-L213

Reply:
and this:

https://github.com/p2panda/p2panda/blob/010416d386f0bb7d94823264c14f4abd437768a8/p2panda-rs/src/encoder.rs#L185-L199

https://github.com/p2panda/p2panda/blob/7435545d214e99c60b928c4bd6393be10ea2fe1d/p2panda-rs/src/encoder.rs#L233-L287